### DISCUSSION TITLE
[Potential Bug] Type-narrowing based on `x in y`

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pyright#9337] [Q&A] [answered]
I'm asking if the current behavior is intentional, because it looks much like a bug to me.

We can now have following snippet pass type-checking -

```python
from typing_extensions import assert_type

def foo(x: float = 0.0, y: list[int] = [0]):
    if x in y:
        _ = assert_type(x, "int")  # !!!
```

The type-narrowing of `x` can be a false negative, as the default value `0.0` of `x` is obviously not an instance of `int`.

On the other hand, type-narrowing based on `x == L` seems to work soundly -

```python
from typing_extensions import Literal, assert_type

def foo(x0: float = 0.0, x1: int = 0, L: Literal[0] = 0):
    if x0 == L:
        assert_type(x0, "float")  # No type-narrowing.
    if x1 == L:
        assert_type(x1, "Literal[0]")  # Safe.
```

My suggestion is that type-narrowing based of `x in y` only take effect when the element type of `y` is a literal type (or maybe a union of literal types that shares a same runtime type) and type of `x` is the runtime type of `y`(or some related union types).

My expected behavior would be like following examples -

```python
from typing_extensions import Literal, assert_type

def f0(x: int, y: list[Literal[0, 1]]):
    if x in y:
        _ = assert_type(x, "Literal[0, 1]")  # Narrowed.

def f1(x: Literal[-1, 0], y: list[Literal[0, 1]]):
    if x in y:
        _ = assert_type(x, "Literal[0]")  # Narrowed.

def f2(x: Literal[0, 1, 2], y: list[Literal[0, 1]]):
    if x in y:
        _ = assert_type(x, "Literal[0, 1]")  # Narrowed.

def f3(x: float, y: list[int]):
    if x in y:
        _ = assert_type(x, "float")  # Not narrowed, because `int` is not a literal type or a union of literal types.

def f4(x: float, y: list[Literal[0, 1]]):
    if x in y:
        _ = assert_type(x, "float")  # Not narrowed, because `float` is not `int`.

def f5(x: int, y: list[Literal[0, True]]):
    if x in y:
        _ = assert_type(x, "int")  # Not narrowed, because `Literal[0]` and `Literal[True]` don't share a same runtime type.

def f6(x: bool, y: list[Literal[0, 1]]):
    if x in y:  # Assuming this is allowed.
        _ = assert_type(x, "bool")  # Not narrowed, because `bool` is not `int`.
```

**Comment by erictraut** [comment, marked as answer]
I agree this is a bug. Thanks for reporting it. I'm surprised no one has reported this previously since it has existed for more than two years. 

Converting to an issue so it is tracked.

**Reply by erictraut** [reply]
Looks like GitHub's "convert to issue" feature is currently broken. @Azureblade3808, could you please file a bug report. That way, it will be filed under your name.

**Reply by Azureblade3808** [reply]
Done as #9338 .
